File: app/src/main/java/com/arkapp/partyplanner/ui/catererList/CaterersListAdapter.java
package com.arkapp.partyplanner.ui.catererList;

import android.annotation.SuppressLint;
import android.app.Activity;
import android.os.AsyncTask;
import android.view.LayoutInflater;
import android.view.ViewGroup;

import androidx.databinding.DataBindingUtil;
import androidx.navigation.ActivityKt;
import androidx.navigation.NavController;
import androidx.recyclerview.widget.RecyclerView;
import androidx.recyclerview.widget.RecyclerView.ViewHolder;

import com.arkapp.partyplanner.R;
import com.arkapp.partyplanner.data.models.Caterer;
import com.arkapp.partyplanner.data.models.PartyDetails;
import com.arkapp.partyplanner.data.repository.PrefRepository;
import com.arkapp.partyplanner.data.room.AppDatabase.Companion;
import com.arkapp.partyplanner.data.room.SummaryDetailsDao;
import com.arkapp.partyplanner.databinding.RvCatererBinding;
import com.arkapp.partyplanner.utils.AddUnfinishedAsyncTask;
import com.arkapp.partyplanner.utils.AppDataKt;
import com.google.gson.Gson;
import com.google.gson.reflect.TypeToken;

import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

import java.lang.reflect.Type;
import java.util.ArrayList;
import java.util.List;

public final class CaterersListAdapter extends RecyclerView.Adapter<RecyclerView.ViewHolder> {
    private final Activity context;
    private final List<Caterer> caterersList;
    private final NavController navController;
    private final PrefRepository prefRepository;

    private final Gson gson = new Gson();
    private final Type type = new TypeToken<ArrayList<String>>() {
    }.getType();

    public CaterersListAdapter(@NotNull Activity context, @NotNull List<Caterer> caterersList, @NotNull NavController navController, @NotNull PrefRepository prefRepository) {
        this.context = context;
        this.caterersList = caterersList;
        this.navController = navController;
        this.prefRepository = prefRepository;
    }

    @NotNull
    public ViewHolder onCreateViewHolder(@NotNull ViewGroup parent, int viewType) {
        return (ViewHolder) (new CaterersListViewHolder((RvCatererBinding)
                DataBindingUtil
                        .inflate(
                                LayoutInflater.from(parent.getContext()),
                                R.layout.rv_caterer,
                                parent,
                                false)));
    }

    @SuppressLint({"ClickableViewAccessibility", "SetTextI18n"})
    public void onBindViewHolder(@NotNull ViewHolder holder, int position) {
        RvCatererBinding binding = ((CaterersListViewHolder) holder).getViewBinding();
        final Caterer caterersData = (Caterer) this.caterersList.get(position);

        binding.name.setText(caterersData.getName().trim());
        binding.price.setText("$" + caterersData.getPricePerPax());
        binding.totalGuestPriceTv.setText(prefRepository.getCurrentPartyDetails().getPartyGuest() + " Pax total");
        binding.totalGuestPrice.setText("$" + (caterersData.getPricePerPax() * prefRepository.getCurrentPartyDetails().getPartyGuest()));

        //This will show all the supported party type of the caterer
        ArrayList<String> partyTypes = gson.fromJson(caterersData.getPartyType(), type);
        StringBuilder partyTypeString = new StringBuilder();

        for (String x : partyTypes) {
            partyTypeString.append(x + ", ");
        }
        binding.partyTypeValue.setText(partyTypeString.toString().substring(0, partyTypeString.toString().length() - 2));

        //On clicking the caterer it open the different screen according to party detail
        binding.parent.setOnClickListener(v -> {
            PartyDetails details = prefRepository.getCurrentPartyDetails();
            details.setSelectedCaterer(caterersData);
            prefRepository.setCurrentPartyDetails(details);

            if (prefRepository.getCurrentPartyDetails().getPartyDestination().equals(context.getString(R.string.home)))
                navController.navigate(R.id.action_caterersListFragment_to_specialSelectionFragment);

            else if (prefRepository.getCurrentPartyDetails().getCheckedItemList() != null &&
                    !prefRepository.getCurrentPartyDetails().getCheckedItemList().isEmpty())
                new UpdateSummaryAsyncTask(context, prefRepository).execute();
            else {
                navController.navigate(R.id.action_caterersListFragment_to_venueLocationFragment);
                new AddUnfinishedAsyncTask(context, prefRepository).execute();
            }
        });
    }

    public int getItemCount() {
        return this.caterersList.size();
    }

    public long getItemId(int position) {
        return (long) (caterersList.get(position)).hashCode();
    }

    private static final class UpdateSummaryAsyncTask extends AsyncTask<Void, Void, Void> {
        private final Activity context;
        private final PrefRepository prefRepository;

        public UpdateSummaryAsyncTask(@NotNull Activity context, @NotNull PrefRepository prefRepository) {
            this.context = context;
            this.prefRepository = prefRepository;
        }

        @Override
        protected Void doInBackground(@NotNull Void... params) {
            SummaryDetailsDao summaryDao = (new Companion()).getDatabase(context).summaryDao();
            summaryDao.delete(prefRepository.getCurrentUser().getUid());
            summaryDao.insert(AppDataKt.convertSummary(
                    prefRepository.getCurrentPartyDetails(),
                    prefRepository.getCurrentUser().getUid()));
            return null;
        }

        @SuppressLint({"SetTextI18n"})
        @Override
        protected void onPostExecute(@Nullable Void summaryData) {
            ActivityKt.findNavController(context, R.id.fragment).navigate(R.id.action_caterersListFragment_to_finalChecklistFragment);
        }
    }
}

File: app/src/main/java/com/arkapp/partyplanner/ui/catererList/CaterersListAdapter.kt
package com.arkapp.partyplanner.ui.catererList

import android.annotation.SuppressLint
import android.app.Activity
import android.os.AsyncTask
import android.view.LayoutInflater
import android.view.ViewGroup
import androidx.databinding.DataBindingUtil
import androidx.lifecycle.LifecycleCoroutineScope
import androidx.navigation.NavController
import androidx.navigation.findNavController
import androidx.recyclerview.widget.RecyclerView
import com.arkapp.partyplanner.R
import com.arkapp.partyplanner.data.models.Caterer
import com.arkapp.partyplanner.data.repository.PrefRepository
import com.arkapp.partyplanner.data.room.AppDatabase
import com.arkapp.partyplanner.utils.AddUnfinishedAsyncTask
import com.arkapp.partyplanner.utils.convertSummary
import com.google.gson.Gson
import com.google.gson.reflect.TypeToken

/**
 * Created by Abdul Rehman on 28-02-2020.
 * Contact email - abdulrehman0796@gmail.com
 */


/**
 * This is a recycler view adapter used to show caterers list
 * */
class CaterersListAdapter(
    private val context: Activity,
    private val caterersList: List<Caterer>,
    private val navController: NavController,
    private val prefRepository: PrefRepository,
    private val lifecycleScope: LifecycleCoroutineScope
) :
    RecyclerView.Adapter<RecyclerView.ViewHolder>() {

    private val gson = Gson()
    private val type = object : TypeToken<ArrayList<String>>() {}.type!!

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): RecyclerView.ViewHolder {
        return CatereresListViewHolder(
            DataBindingUtil.inflate(
                LayoutInflater.from(parent.context),
                R.layout.rv_caterer,
                parent,
                false
            )
        )
    }

    @SuppressLint("ClickableViewAccessibility", "SetTextI18n")
    override fun onBindViewHolder(holder: RecyclerView.ViewHolder, position: Int) {
        val binding = (holder as CatereresListViewHolder).viewBinding

        val caterersData = caterersList[position]
        binding.name.text = caterersData.name.trim()
        binding.price.text = "$${caterersData.pricePerPax}"
        binding.totalGuestPriceTv.text = "${prefRepository.currentPartyDetails.partyGuest} Pax total"
        binding.totalGuestPrice.text = "$${caterersData.pricePerPax * prefRepository.currentPartyDetails.partyGuest!!}"

        //This will show all the supported party type of the caterer
        val partyTypes = gson.fromJson<ArrayList<String>>(caterersData.partyType, type)
        var partyTypeString = ""

        for (x in partyTypes) {
            partyTypeString += "$x, "
        }
        binding.partyTypeValue.text = partyTypeString.substring(0, partyTypeString.length - 2)

        //On clicking the caterer it open the different screen according to party detail
        binding.parent.setOnClickListener {
            val details = prefRepository.currentPartyDetails
            details.selectedCaterer = caterersData
            prefRepository.currentPartyDetails = details

            when {
                prefRepository.currentPartyDetails.partyDestination == context.getString(R.string.home) ->
                    navController.navigate(R.id.action_caterersListFragment_to_specialSelectionFragment)
                !prefRepository.currentPartyDetails.checkedItemList.isNullOrEmpty() ->
                    UpdateSummaryAsyncTask(context, prefRepository).execute()
                else -> {
                    navController.navigate(R.id.action_caterersListFragment_to_venueLocationFragment)
                    AddUnfinishedAsyncTask(context, prefRepository).execute()
                }
            }
        }

    }


    override fun getItemCount() = caterersList.size

    override fun getItemId(position: Int): Long {
        return caterersList[position].hashCode().toLong()
    }

    // Used to update the summary data in SQL table
    private class UpdateSummaryAsyncTask(private val context: Activity,
                                         private val prefRepository: PrefRepository) : AsyncTask<Void, Void, Void?>() {

        override fun doInBackground(vararg params: Void?): Void? {
            val summaryDao = AppDatabase.Companion().getDatabase(context).summaryDao()
            summaryDao.delete(prefRepository.currentUser?.uid!!)
            summaryDao.insert(convertSummary(prefRepository.currentPartyDetails,
                                             prefRepository.currentUser?.uid!!))
            return null
        }

        @SuppressLint("SetTextI18n")
        override fun onPostExecute(summaryData: Void?) {
            context.findNavController(R.id.fragment)
                .navigate(R.id.action_caterersListFragment_to_finalChecklistFragment)
        }
    }

}